services/auth/src/signaturevalidator.js

```diff
@@ -234,14 +234,23 @@ const createSignatureValidator = function(options) {
     let accessToken, scopes, expires;
     ({ clientId, expires, accessToken, scopes } = await options.clientLoader(issuingClientId));
 
+    // Implicitly grant all clients the anonymous role _before_ we check
+    // `authorizedScopes`. If we didn't do this first then the check that
+    // the client has a superset of `authorizedScopes` can fail if the scope
+    // is provided by it being in `anonymous`
+    scopes = utils.mergeScopeSets(scopes, ['assume:anonymous']);
+    scopes = options.expandScopes(scopes);
+
```

Comment:
These new lines (237-243) can be moved inside the `if (ext) {` condition below, so that the merge/expand only happens twice if there is an `ext`.

An alternative solution would be for the `utils.satisfiesExpression` function to consider anonymous scopes _only if satisfaction fails without them_. This would probably have some performance benefit and hopefully isolate the scope satisfaction code to a single place, especially if e.g. workers are using the same validation libraries. However, this injects the concept of anonymous scopes at a pretty low level of the conceptual stack, and there is an argument that it is a higher level concern whether a request carries anonymous scopes or not, and shouldn't be considered at a lexical evaluation level. Let's discuss this afternoon.

Reply:
I agree that we can avoid doing this twice.

This is already really difficult to reason about, and that's *with* a firm definition of "satisfaction" that does not involve role expansion. I think blurring those two things will make this even harder to reason about.